src/Sarif.Multitool/Rules/SARIF1017.InvalidIndex.cs

```diff
@@ -0,0 +1,209 @@
+    {
+        private readonly MultiformatMessageString _fullDescription = new MultiformatMessageString
+        {
+            Text = RuleResources.SARIF1017_InvalidIndex
+        };
+
+        public override MultiformatMessageString FullDescription => _fullDescription;
+
+        public override FailureLevel DefaultLevel => FailureLevel.Error;
+
+        /// <summary>
+        /// SARIF1017
+        /// </summary>
+        public override string Id => RuleId.InvalidIndex;
+
+        protected override IEnumerable<string> MessageResourceNames => new string[]
+        {
+            nameof(RuleResources.SARIF1017_Default)
+        };
+
+        protected override void Analyze(Address address, string addressPointer)
+        {
+            ValidateArrayIndex(
+                address.Index,
+                Context.CurrentRun.Addresses,
+                addressPointer,
+                "address",
+                "index",
+                $"runs[{Context.CurrentRunIndex}].addresses");
+
+            ValidateArrayIndex(
+                address.ParentIndex,
+                Context.CurrentRun.Addresses,
+                addressPointer,
+                "address",
+                "parentIndex",
+                $"runs[{Context.CurrentRunIndex}].addresses");
+        }
+
+        protected override void Analyze(Artifact artifact, string artifactPointer)
+        {
+            ValidateArrayIndex(
+                artifact.ParentIndex,
+                Context.CurrentRun.Artifacts,
+                artifactPointer,
+                "artifact",
+                "parentIndex",
+                $"runs[{Context.CurrentRunIndex}].artifacts");
+        }
+
+        protected override void Analyze(ArtifactLocation artifactLocation, string artifactLocationPointer)
+        {
+            ValidateArrayIndex(
+                artifactLocation.Index,
+                Context.CurrentRun.Artifacts,
+                artifactLocationPointer,
+                "artifactLocation",
+                "index",
+                $"runs[{Context.CurrentRunIndex}].artifacts");
+        }
+
+        protected override void Analyze(GraphTraversal graphTraversal, string graphTraversalPointer)
+        {
+            ValidateArrayIndex(
+                graphTraversal.RunGraphIndex,
+                Context.CurrentRun.Graphs,
+                graphTraversalPointer,
+                "graphTraversal",
+                "runGraphIndex",
+                $"runs[{Context.CurrentRunIndex}].graphTraversals");
+
+            ValidateArrayIndex(
+                graphTraversal.ResultGraphIndex,
+                Context.CurrentResult.Graphs,
+                graphTraversalPointer,
+                "graphTraversal",
+                "resultGraphIndex",
+                $"runs[{Context.CurrentRunIndex}].results[{Context.CurrentResultIndex}].graphTraversals");
+        }
+
+        protected override void Analyze(LogicalLocation logicalLocation, string logicalLocationPointer)
```

Comment:
>LogicalLocation [](start = 40, length = 15)

where's the check for logicalLocation.parentIndex?

Reply:
I have not finished all the cases yet. And some of the cases have sub-cases: for instance, you have to validate `ReportingDescriptorReference.Index`, but `ReportingDescriptorReferences` occur in several different contexts -- and you need tests for all of them to ensure you've got error message arguments right in each case.

---
In reply to: [305428389](https://github.com/microsoft/sarif-sdk/pull/1570#discussion_r305428389) [](ancestors = 305428389)

Reply:
Having said that, I just looked at my checklist (I pasted it into another comment), and I see that I put an `x` in the `logicalLocation.parentIndex` box -- and that was a mistake. Good catch! I'll uncheck it and implement it (and then re-check it ;-)).

---
In reply to: [305525258](https://github.com/microsoft/sarif-sdk/pull/1570#discussion_r305525258) [](ancestors = 305525258,305428389)